config/crds/kubic_v1beta1_dexconfiguration.yaml

```diff
@@ -0,0 +1,88 @@
+apiVersion: apiextensions.k8s.io/v1beta1
```

Comment:
yop agree on this

Reply:
Integration tests that were on the repo are going to load them.

- https://github.com/kubic-project/dex-operator/blob/master/pkg/controller/dex/dex_controller_suite_test.go#L43-L45
- https://github.com/kubic-project/dex-operator/blob/master/pkg/apis/kubic/v1beta1/v1beta1_suite_test.go#L41-L43

Without loading these CRD's on the integration `apiserver` they cannot be found, and integration tests fail.

Reply:
I think we have two options:

* to generate those files before running the integration tests, as part of the integration tests build.
* to create them in advance and add them as `fixtures` (as @MalloZup mentioned)

I would prefer the first solution, as I think creating them in advance is not a good idea as it could differ from the code that is being tested. In conclusion, I would create a `make test-integration` (or something like that), and that target could depend on `generate-crds`.

Reply:
The default project created by `kubebuilder` generates those files here; so I don't see a strong reason to change the location of these files. I see that if we need to deviate from `kubebuilder` for a good reason it's fine, but the more close we are to it the better, IMHO.

Reply:
> I think creating them in advance is not a good idea as it could differ from the code that is being tested

They would generated when you build the project. These directories are by default not included in the `.gitignore` file with `kubebuilder`. Do we really want to deviate from `kubebuilder` scaffold right now in this trivial thing?

I can see the benefit in having the CRD's YAML's being regenerated everytime you build. In this regard all projects that used `kubebuilder` as a scaffolding tool would behave the same.

I don't have issues deviating from `kubebuilder`, what what's the real benefit in changing this right now?

Reply:
https://github.com/kubernetes-sigs/kubebuilder/blob/master/docs/book/basics/project_creation_and_structure.md#config

Reply:
yop agreed on that to stay true the upstream framework. 

If this files are auto-generated, i would add a gitignore to prevent commit of this and i would be in favor to don't put them here and having a makefile target for that sounds good to me :lollipop: :tropical_fish:

Reply:
The problem is we are not distributing those files but the `deployment/*.yaml` file (because that file contains everything needed by users). In my opinion, adding those file to the repo would be confusing for users.

Besides that, the `kubebuilder` docs do not say anything about commiting those files to the repo...

Reply:
> Besides that, the kubebuilder docs do not say anything about commiting those files to the repo...

They don't, but they generate the files and they include a `.gitignore` file that by default does not include `config/{crds,rbac}`... I'd assume that means that they are not meant to be ignored by git, as they are ignoring other directories/files by default.

I'd like to deviate the minimum as possible from the scaffold where possible and only when we have reasons to do so, since it will be easier for external contributors to help if we are closer to `kubebuilder` than something completely different. I also am a bit hesitant of including a magic `Makefile` that does many things.

I'm still having a look at this though, and I have asked about the general approach regarding git and `config/{crds,rbac}` on the kubebuilder channel in Kubernetes' slack.

Reply:
> I am trying to figure out why we deviated this much from the scaffold and got to a much more complex Makefile that is partially replacing some code generation.

Well, I think we did not _"deviate"_ but we evolved the `Makefile`, and this one can do many more things than the minimal scaffold Makefile that `kubebuilder` generates. I don't see any problem on adding more features to the Makefile...

Reply:
Targets have been renamed, old targets don't exist anymore. We are using `kustomize` in a slightly different way than upstream, still unclear why it's better for us (we are not saving config/{crds,rbac} but persisting the `kustomize` output and removing the old ones, while the default project creates the kustomize output on the fly and `kubectl apply -f -` it).

To be honest, I see lots of changes that don't add much value, but that make the project hard to start with if you already interacted with a `kubebuilder` scaffold before. These projects did not grow incredibly complex yet, and I wonder if we really need all this `Makefile` changes [the interface to the developer] right now.

I mean, the original `Makefile` looks incredibly clean, do we need to add so many modifications to it?

I know it might sound like I'm being incredibly picky, but I think right now is the best time to set a standard that we can follow on all the operators, and starting all operators from a deviation of the original `kubebuilder` scaffold Makefile with little gain worries me a bit.

Reply:
I started with the standard Makefile created by the `kubebuilder` and it started evolving, reaching this point after adding many features that I needed. I think scaffold code is never intended to remain as it is, so it evolved. And I don't think a `Makefile` is an entripoint for a project for anybody (otherwise, people would never contribute to the kernel :blush:), so I don't really think people will be scared of this...

Anyway, if you want to start a Makefile from scratch please go ahead. I think you will probably have to re-add some of the features that I added, but any cleanup will be wellcome.

Reply:
I absolutely understand your points. And I thank you for the (very much needed) initial work on all this.

I hate to look like I want to dismiss work that is already done, but myself and others found issues with the building of the projects (I found some typos on the Makefile in which I lost some time, in tasks that were supposed to be usually run, like `make test`).

My main issue with the current Makefiles is that maybe they are more complicated than they should. I know this sounds stupid. It's just a Makefile. Since we have two operators and maybe we'll have more in the future I think it would be a good idea to try to follow the same conventions on all of them. The easier the conventions, the easier to maintain these conventions across all operators (whenever possible).

>  And I don't think a Makefile is an entripoint for a project for anybody (otherwise, people would never contribute to the kernel blush), so I don't really think people will be scared of this...

Aside from the joke, I think you are right here :+1:, with a good documentation it doesn't matter how many targets or how complex the Makefile is. However I think it would be good if we could stop for a second, and set the standards for the two operators that we have and then extend what we can do with them, and be consistent across operators on how we test/build and develop on them.

Reply:
I definitely agree that we should not be scared of adding new functionality, but on top of the standard make file as much as possible. For example, renaming targets doesn't seam to me like required and makes understanding the project more complex, as it deviates from the standard proposed by the framework.